Move polling to azure-core

bryevdv opened the pull request:
This PR migrates poling support from msrest to azure-core

* Use/expect `HttpResponse` throughout (no more `ClientRawResponse`)
* Use/expect `PipelineClient` thoughtout (no more `ServiceClient`)

It seems as though some of the type specifications can be tightened up now that the conditional switching on `ClientRawResponse` and `ServiceClient` is removed. @annatisch noted that there will likely need to be an async version of `PipelineClient` so there will likely be some small changes from that as well. 

cc @annatisch

adxsdk6 commented:
Can one of the admins verify this patch?

adxsdk6 commented:
Can one of the admins verify this patch?

annatisch commented on azure-core/azure/core/polling/async_poller.py:
```diff
@@ -0,0 +1,80 @@
+    """
+
+    if isinstance(deserialization_callback, type) and issubclass(deserialization_callback, Model):
+        deserialization_callback = deserialization_callback.deserialize
```
I wonder if there's a way we can remove the dependency on `Model` here - for example:
```python
try:
    deserialization_callback = deserialization_callback.deserialize
except AttributeError:
    pass
```
I don't know if this would actually work - but the autorest tests should tell us

annatisch commented on azure-core/azure/core/polling/poller.py:
```diff
@@ -0,0 +1,233 @@
+        self._polling_method = polling_method
+
+        if isinstance(deserialization_callback, type) and issubclass(deserialization_callback, Model):
+            deserialization_callback = deserialization_callback.deserialize  # type: ignore
```
Same comment as above re Model

bryevdv commented:
@annatisch I addressed your comments in https://github.com/Azure/azure-sdk-for-python/pull/4888/commits/df0f7a647d80f6c17cf38d789d9c846f48ee492a Note that I still included a conditional import of `Model` for `TYPE_CHECKING` since it appears to still be needed in that instance
